### DISCUSSION TITLE
Downgrading

### DISCUSSION THREAD

**Discussion by marian-m12l** [marian-m12l/studio#145] [General]
Making this thread to continue the conversation started on #122 about downgrading to firmware 1.x. Please **do not try this at home** :-D (unless you know what you're doing... somewhat), downgrading is still _terra incognita_ and can brick the device, and **definitely** voids your warranty.

**Comment by noctanvol** [comment]
> As of today my Daugther won't stop using her lunii sweat_smile
> 
> ah ah, yeah, I know this "problem" :-D
> 
> Did you give it a try with version 0.2.0? It should at least give another error, might be helpful. Btw, on what OS are you running STUdio?

**YES I did try the latest version as well ( i'll send you the logs later today)**

> If I understand correctly, you're using STUdio without error on the "donor" device?

**yes on the donor lunii everything work perfectly whatever the studio version !**

> What SD card are you using in the end? 8GB or 32 GB? I only have a 4GB version of the Lunii, which relies on a hard-wired card size in the driver, so there may be some bugs with bigger cards that I don't know of. Is the "donor" card the same size as the receiver?

**I'm currently using the original lunii  8gb SD card ( i wanted to save it in case of rollback was needed but i ran out of option so I put her back in after cloning the donor lunii)**

**By the way this is really weird that you have on old lunii (with 4gb) with the same motherboard layout as my own (8gb)
in both our case the chip contaning the firmare was on the screen side wheras on my donor lunii the chip was directly accessible...**

**I'll post photo today if I can**

**regards**

**Reply by noctanvol** [reply]
BE CARREFUL IF YOU WANT TO TRY DOWNGRADING your LUNII, you might brick you device and WILL void your warranty... ( mine was out of warranty anyway...) 
**What I did is really a cloning of device rather than a real downgrade** ( tough I did end up with an older version of the firmwareI originally had, **for now I lost my lunii's IUD in the process**)
bellow is what I did and not everything is working for me as of today. **I had backup **( from the old firmware) of all my story( even the one I bought !) so I was able to get them back after.
**After the downgrade/cloning you might not be able to download them from the luniistore anymore !** ( you basically get your lunii's donor UID not your own)

This is the tools what i used for downgrading
![IMG_20201208_141451](https://user-images.githubusercontent.com/64204617/101607275-b1ab3600-3a04-11eb-82ba-54990a5d2141.jpg)
![IMG_20201208_141418 - Copie](https://user-images.githubusercontent.com/64204617/101613302-e4a4f800-3a0b-11eb-9cfd-da9223ce4b06.jpg)
you have to be careful that all pins are connected, do it one side at a time, the clip is really handy there.
The red wire on the clip MUST be connected to the the pin where you have the tiny hole on the chip ( i put a red circle on it in the next pictures)

**HARDWARE**
 SOIC8 SOP8 pince puce + programer WCH341A (mine is pro, not sure it changes anything...) 
I used a linux live usb key to perform the reading and flashing of the chips ( I dont really trust the drivers and software i found on windows) and linux does it almost natively with "flashrom"

command shell linux
**_lsusb_**
to check if the usbkey is detected
![Screenshot_20201208_131730](https://user-images.githubusercontent.com/64204617/101617796-60557380-3a11-11eb-92db-aff803210c27.png)


to perform a read on the chip connected
**_sudo flashrom -p ch341a_spi -r mon_fichier.bin_**
-p stand pour "programmer" -r for "read"

to perform a write figure it out using "man flashrom"  :) **### (BE VERY CAREFUL ON WHAT YOU DO !!!)**

this video helped me a lot for using the programmer
https://www.youtube.com/watch?v=OttN8wQ1HhE

for cloning the SD card i used Win32 Disk Imager : https://sourceforge.net/projects/win32diskimager/

here is my donor lunii
![IMG_20201208_135858 - Copie](https://user-images.githubusercontent.com/64204617/101608722-5ed27e00-3a06-11eb-8a55-a5cc707d1b55.jpg)
![IMG_20201208_135907 - Copie](https://user-images.githubusercontent.com/64204617/101608731-609c4180-3a06-11eb-94c4-daa5f88d2da3.jpg)
![IMG_20201208_140358 - Copie](https://user-images.githubusercontent.com/64204617/101613468-15852d00-3a0c-11eb-8a41-c23c1bb400b5.jpg)


Here is my own
![IMG_20201208_135839 - Copie](https://user-images.githubusercontent.com/64204617/101608948-a9ec9100-3a06-11eb-908e-290682084ec7.jpg)
![IMG_20201208_135852 - Copie](https://user-images.githubusercontent.com/64204617/101608954-ab1dbe00-3a06-11eb-950e-41e6233e8982.jpg)
![IMG_20201205_160253 - Copie](https://user-images.githubusercontent.com/64204617/101613549-2cc41a80-3a0c-11eb-81d7-bc3ed33d95c9.jpg)



sorry for the huge images... ( I croped some...) :)

**Reply by noctanvol** [reply]
@marian-m12l 
>Did you give it a try with version 0.2.0? It should at least give another error, might be helpful. Btw, on what OS are you running STUdio?

I tried with studio 2.0, i'm using Win 10 home x64

here are the logs from studio

_2020-12-08 19:59:58,017 [INFO ] [studio.webui.service.StoryTellerService] [StoryTellerService.java:58]: Device plugged
2020-12-08 20:00:10,634 [WARN ] [studio.webui.service.LibraryService] [LibraryService.java:122]: Pack to transfer is in archive format. Converting to binary format and storing in temporary file: C:\Users\NOCTAN~1\AppData\Local\Temp\Casse-noisette-b111eefb-467d-4a48-a9ce-8795f5771272-v1.zip7706117227726872782.pack
2020-12-08 20:00:10,635 [WARN ] [studio.webui.service.LibraryService] [LibraryService.java:124]: Reading archive format pack
2020-12-08 20:00:10,747 [WARN ] [studio.webui.service.LibraryService] [LibraryService.java:133]: Uncompressing pack assets
2020-12-08 20:00:19,810 [WARN ] [studio.webui.service.LibraryService] [LibraryService.java:137]: Writing binary format pack
2020-12-08 20:00:20,843 [INFO ] [studio.webui.service.StoryTellerService] [StoryTellerService.java:141]: Transferring pack to device: 101361152 bytes
2020-12-08 20:00:20,844 [INFO ] [studio.webui.service.StoryTellerService] [StoryTellerService.java:143]: Transferring pack to device: 197971 sectors
2020-12-08 20:00:25,888 [ERROR] [studio.driver.LibUsbHelper] [LibUsbHelper.java:391]: **Read operation failed while writing to SD**
2020-12-08 20:00:25,889 [ERROR] [studio.webui.service.StoryTellerService] [StoryTellerService.java:165]: Failed to add pack to device
java.util.concurrent.CompletionException: studio.driver.StoryTellerException: Read operation failed while writing to SD
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.base/java.util.concurrent.CompletableFuture.complete(Unknown Source)
	at studio.driver.LibUsbHelper.lambda$asyncTransferIn$17(LibUsbHelper.java:446)
	at org.usb4java.LibUsb.handleEventsTimeout(Native Method)
	at studio.driver.LibUsbAsyncEventsWorker.run(LibUsbAsyncEventsWorker.java:34)
Caused by: studio.driver.StoryTellerException: Read operation failed while writing to SD
	at studio.driver.LibUsbHelper.lambda$asyncWriteSDSectors$13(LibUsbHelper.java:392)
	... 6 more
2020-12-08 20:01:08,029 [INFO ] [studio.driver.LibUsbActivePollingWorker] [LibUsbActivePollingWorker.java:55]: Active polling lost the device. Firing event.
2020-12-08 20:01:08,029 [INFO ] [studio.webui.service.StoryTellerService] [StoryTellerService.java:88]: Device unplugged_

I always get this error whatever version of studio I use...

**Reply by GotoHack** [reply]
> BE CARREFUL IF YOU WANT TO TRY DOWNGRADING your LUNII, you might brick you device and WILL void your warranty... ( mine was out of warranty anyway...)
> **What I did is really a cloning of device rather than a real downgrade** ( tough I did end up with an older version of the firmwareI originally had, **for now I lost my lunii's IUD in the process**)
> bellow is what I did and not everything is working for me as of today. **I had backup **( from the old firmware) of all my story( even the one I bought !) so I was able to get them back after.
> **After the downgrade/cloning you might not be able to download them from the luniistore anymore !** ( you basically get your lunii's donor UID not your own)
> 
> This is the tools what i used for downgrading
> ![IMG_20201208_141451](https://user-images.githubusercontent.com/64204617/101607275-b1ab3600-3a04-11eb-82ba-54990a5d2141.jpg)
> ![IMG_20201208_141418 - Copie](https://user-images.githubusercontent.com/64204617/101613302-e4a4f800-3a0b-11eb-9cfd-da9223ce4b06.jpg)
> you have to be careful that all pins are connected, do it one side at a time, the clip is really handy there.
> The red wire on the clip MUST be connected to the the pin where you have the tiny hole on the chip ( i put a red circle on it in the next pictures)
> 
> **HARDWARE**
> SOIC8 SOP8 pince puce + programer WCH341A (mine is pro, not sure it changes anything...)
> I used a linux live usb key to perform the reading and flashing of the chips ( I dont really trust the drivers and software i found on windows) and linux does it almost natively with "flashrom"
> 
> command shell linux
> **_lsusb_**
> to check if the usbkey is detected
> ![Screenshot_20201208_131730](https://user-images.githubusercontent.com/64204617/101617796-60557380-3a11-11eb-92db-aff803210c27.png)
> 
> to perform a read on the chip connected
> **_sudo flashrom -p ch341a_spi -r mon_fichier.bin_**
> -p stand pour "programmer" -r for "read"
> 
> to perform a write figure it out using "man flashrom" :) **### (BE VERY CAREFUL ON WHAT YOU DO !!!)**
> 
> this video helped me a lot for using the programmer
> https://www.youtube.com/watch?v=OttN8wQ1HhE
> 
> for cloning the SD card i used Win32 Disk Imager : https://sourceforge.net/projects/win32diskimager/
> 
> here is my donor lunii
> ![IMG_20201208_135858 - Copie](https://user-images.githubusercontent.com/64204617/101608722-5ed27e00-3a06-11eb-8a55-a5cc707d1b55.jpg)
> ![IMG_20201208_135907 - Copie](https://user-images.githubusercontent.com/64204617/101608731-609c4180-3a06-11eb-94c4-daa5f88d2da3.jpg)
> ![IMG_20201208_140358 - Copie](https://user-images.githubusercontent.com/64204617/101613468-15852d00-3a0c-11eb-8a41-c23c1bb400b5.jpg)
> 
> Here is my own
> ![IMG_20201208_135839 - Copie](https://user-images.githubusercontent.com/64204617/101608948-a9ec9100-3a06-11eb-908e-290682084ec7.jpg)
> ![IMG_20201208_135852 - Copie](https://user-images.githubusercontent.com/64204617/101608954-ab1dbe00-3a06-11eb-950e-41e6233e8982.jpg)
> ![IMG_20201205_160253 - Copie](https://user-images.githubusercontent.com/64204617/101613549-2cc41a80-3a0c-11eb-81d7-bc3ed33d95c9.jpg)
> 
> sorry for the huge images... ( I croped some...) :)

Hello, I had some trouble with the 2.4 update. Now I have a working device by I lost the stories. I did a backup of my sdcard using dd before the update. I would love to downgrade to the old fw. Is there any chance that you could share your donnor spi dump (I have the same board) ? 
Is it possible to extract the stories from my dump and convert them to the new format and upload them back to the device ?

**Reply by noctanvol** [reply]
> 
> Hello, I had some trouble with the 2.4 update. Now I have a working device by I lost the stories. I did a backup of my sdcard using dd before the update. I would love to downgrade to the old fw. Is there any chance that you could share your donnor spi dump (I have the same board) ?
> Is it possible to extract the stories from my dump and convert them to the new format and upload them back to the device ?
@GotoHack 
hi 

I'm sory but I can't send you the old firware, as it not mine to give. plus the FW include the ID linked to the original account. it might create a mess for the original donor.... 

do you have your customs stories original files  ? then you might be able to convert them to the new format. I didn't check or tested for a long time on the newest version of studio but i think it does what you want now. maybe other users can help you with the convertion.
regards

**Comment by marian-m12l** [comment]
@noctanvol Actually I think the high res photos of the PCB are good to have. Looks like there are plenty of different revisions of this board. On mine, the microcontroller (under the black blob) is on a small PCB on top of the main one.

How about the size of the donor SD card vs the receiver (8GB) SD card?

Regarding the transfer error, are you able to build from the latest master? I could add a few interesting logs.

**Comment by noctanvol** [comment]
Both SD card are the same size, 8gb.

I'll try to do a full clean install of everything ( studio, Java binary et lunii store) there must be something else that is wrong. It should work...

**Comment by marian-m12l** [comment]
You can get the nightly build [here](https://github.com/marian-m12l/studio/actions/runs/412857217). I added some logs and enabled debug logs for the driver.
Are you testing both devices on the same computer?

**Reply by noctanvol** [reply]
> You can get the nightly build [here](https://github.com/marian-m12l/studio/actions/runs/412857217). I added some logs and enabled debug logs for the driver.
> Are you testing both devices on the same computer?

yes I'm testing both lunii on the same computer, for me they are a perfect software clone ... so it should work, but the hardware beeing different from the donor is probably the reason it fails on my lunii.

I'll try the nightly buid tonight to see if it fixes the write error, 
thx

**Comment by noctanvol** [comment]
same error with the night build, I hope it will help you understand what is failing when writting
[studio-latest.log](https://github.com/marian-m12l/studio/files/5675580/studio-latest.log)
thx

**Comment by marian-m12l** [comment]
Thanks for the logs.
I have a few questions : did this device work previously with STUdio? Did you try several USB cables (and it is the same for the other device)? Did you try transferring from the device to the computer?
From the errors, it looks like the USB transfer is incomplete. I'll have a better look, but you could also try a few tweaks if you're comfortable with building STUdio from the sources?
